lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -0,0 +1,481 @@
+        Sqlite.bind statement params
+        callback (Sqlite.nextRow statement)
+
+------------------------------------------------------------------------------------------------------------------------
+-- Exceptions
+
+-- | A query was expected to return exactly one row, but it did not. The exception carries a string representation of
+-- the rows that were actually returned.
+newtype ExpectedExactlyOneRowException = ExpectedExactlyOneRowException
+  { rows :: String
+  }
+  deriving stock (Show)
+
+-- | A query was expected to return exactly one row, but it did not. The exception carries a string representation of
+-- the rows that were actually returned.
+newtype ExpectedAtMostOneRowException = ExpectedAtMostOneRowException
+  { rows :: String
+  }
+  deriving stock (Show)
+
+data SqliteExceptionInfo a = SqliteExceptionInfo
+  { sql :: Sql,
+    params :: Maybe a,
+    exception :: SomeShowTypeable,
+    connection :: Connection
+  }
+
+throwSqliteException :: SqliteExceptionInfo params -> IO a
+throwSqliteException SqliteExceptionInfo {connection, exception, params, sql} = do
+  threadId <- myThreadId
+  throwIO
+    SqliteException
+      { sql,
+        params = maybe "" anythingToString params,
+        exception,
+        connection,
+        threadId
+      }
+
+data SomeShowTypeable
+  = forall e. (Show e, Typeable e) => SomeShowTypeable e
+
+instance Show SomeShowTypeable where
+  show (SomeShowTypeable x) = show x
+
+-- | A @SqliteException@ represents an exception paired with some context that resulted in the exception.
+--
+-- A @SqliteException@ may result from a number of different conditions:
+--
+-- * The underlying sqlite library threw an exception, as when establishing a connection to a non-existent database.
+-- * A postcondition violation of a function like 'queryMaybe', which asserts that the resulting relation will have
+--   certain number of rows,
+-- * A postcondition violation of a function like 'queryListCheck', which takes a user-defined check as an argument.
+--
+-- A @SqliteException@ should not be inspected or used for control flow when run in a trusted environment, where the
+-- database can be assumed to be uncorrupt. Rather, wherever possible, the user of this library should write code that
+-- is guaranteed not to throw exceptions, by checking the necessary preconditions first. If that is not possible, it
+-- should be considered a bug in this library.
+--
+-- When actions are run on an untrusted codebase, e.g. one downloaded from a remote server, it is sufficient to catch
+-- just one exception type, @SqliteException@.
+data SqliteException = SqliteException
+  { sql :: Sql,
+    params :: String,
+    -- | The inner exception. It is intentionally not 'SomeException', so that calling code cannot accidentally
+    -- 'throwIO' domain-specific exception types, but must instead use a @*Check@ query variant.
+    exception :: SomeShowTypeable,
```

Comment:
I'm not exactly sure what you mean by:

> so that calling code cannot accidentally 'throwIO' domain-specific exception types

Which "Domain-specific" things are you worried about here? Do you mean so that they don't re-throw sqlite exceptions, or something else?

Reply:
The intention is to disallow this kind of thing:

```haskell
result <- query stuff
when (looksWeird result) (throwIO WeirdStuffHappened)
```

because the exception does not contain any context about what query was run, etc.

Instead, you're supposed to use `*Check`:

```haskell
queryCheck stuff (\result -> if looksWeird result then Left WeirdStuffHappened else Right result)
```